In [2]:
import pandas as pd
import numpy as np
tabela_dieta_csv = pd.read_csv('dietaDados.csv')
tabela = tabela_dieta_csv.T
tabela

0                 1   \
Mercadoria            Farinha de trigo (enriquecida)  Farinha de milho   
Calorias(1000)                                  44.7              36.0   
Proteinas(gramas)                             1411.0             897.0   
Cálcio(gramas)                                   2.0               1.7   
Ferro(mg)                                      365.0              99.0   
Vitamina A(1000I.U.)                             0.0              30.9   
Tiamina(mg.)                                    55.4              17.4   
Riboflavina(mg.)                                33.3               7.9   
Niacina(mg.)                                   441.0             106.0   
Ácido Ascórbico(mg.)                             0.0               0.0   

                                          2          3                 4   \
Mercadoria            Leite evaporado (lata)  Margarina  Queijo (cheddar)   
Calorias(1000)                           8.4       20.6               7.4   
Proteinas(gramas)                      422.0       17.0             448.0   
Cálcio(gramas)                          15.1        0.6              16.4   
Ferro(mg)                                9.0        6.0              19.0   
Vitamina A(1000I.U.)                    26.0       55.8              28.1   
Tiamina(mg.)                             3.0        0.2               0.8   
Riboflavina(mg.)                        23.5        0.0              10.3   
Niacina(mg.)                            11.0        0.0               4.0   
Ácido Ascórbico(mg.)                    60.0        0.0               0.0   

                                     5      6             7   \
Mercadoria            Pasta de amendoim  Bacon  Fígado (boi)   
Calorias(1000)                     15.7   41.7           2.2   
Proteinas(gramas)                 661.0    0.0         333.0   
Cálcio(gramas)                      1.0    0.0           0.2   
Ferro(mg)                          48.0    0.0         139.0   
Vitamina A(1000I.U.)                0.0    0.2         169.2   
Tiamina(mg.)                        9.6    0.0           6.4   
Riboflavina(mg.)                    8.1    5.0          50.8   
Niacina(mg.)                      471.0    5.0         316.0   
Ácido Ascórbico(mg.)                0.0    0.0         525.0   

                                         8            9   ...       11  \
Mercadoria            Lombo de porco assado  Salmão Rosa  ...  Repolho   
Calorias(1000)                          4.4          5.8  ...      2.6   
Proteinas(gramas)                     249.0        705.0  ...    125.0   
Cálcio(gramas)                          0.3          6.8  ...      4.0   
Ferro(mg)                              37.0         45.0  ...     36.0   
Vitamina A(1000I.U.)                    0.0          3.5  ...      7.2   
Tiamina(mg.)                           18.2          1.0  ...      9.0   
Riboflavina(mg.)                        3.6          4.9  ...      4.5   
Niacina(mg.)                           79.0        209.0  ...     26.0   
Ácido Ascórbico(mg.)                    0.0          0.0  ...   5369.0   

                          12       13         14           15              16  \
Mercadoria            Cebola  Batatas  Espinafre  Batata0doce  Pêssegos secos   
Calorias(1000)           5.8     14.3        1.1          9.6             8.5   
Proteinas(gramas)      166.0    336.0      106.0        138.0            87.0   
Cálcio(gramas)           3.8      1.8        0.0          2.7             1.7   
Ferro(mg)               59.0    118.0      138.0         54.0           173.0   
Vitamina A(1000I.U.)    16.6      6.7      918.4        290.7            86.8   
Tiamina(mg.)             4.7     29.4        5.7          8.4             1.2   
Riboflavina(mg.)         5.9      7.1       13.8          5.4             4.3   
Niacina(mg.)            21.0    198.0       33.0         83.0            65.0   
Ácido Ascórbico(mg.)  1184.0   2522.0     2755.0       19

### Segundo as iniciais do meu nome, a seguinte distribuição de nutrientes se apresenta:
Para a letra F, teremos 5.9 calorias de restrição, para a letra A, teremos 0.07 gramas de cálcio, 6 unidades internacionais de vitamina A, para a letra I, teremos 6.5 riboflavina e 10 gramas de ácido ascórbico.

Para as demais vitaminas e nutrientes, teremos uma restrição de maior igual a zero.

Para utilizar a rotina linprog da biblioteca SciPy, precisamos nos comportar segundo especificação. Não há inequação de maior ou igual, apenas de menor ou igual e igualdade (a_ub, b_ub, a_eq, b_eq). A primeira dupla representa a matriz e vetor de inequações de restrições e a segundo matriz e vetor de igualdades de restrições.

Com isso, todas as entradas da matriz de restrições e o vetor de restrições precisam ter seu valor invertido devido ao fato de queremos maior ou igual a zero, não o contrário, que é a forma que a rotina espera.

In [3]:
restricoes = np.array([-5.9,0.0,-.07,0.0,-6.0,0.0,-6.5,0.0,-10.0])

Tranformando a tabela em matriz de restrições e transformando os valores em negativo para continuar a implementação da especificação.

In [4]:
matriz = tabela.drop('Mercadoria').to_numpy()
for i in range(len(matriz)):
    matriz[i] = -matriz[i]

Chamando a rotina Linprog da biblioteca SciPy e escolhendo o método simplex para resolver o sistema linear

In [5]:
import numpy as np
from scipy.optimize import linprog

c = np.array([1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1])
A = matriz
B = restricoes
 
res = linprog(c, A_ub=A, b_ub=B,bounds=(0, None), method='simplex') 
res

     con: array([], dtype=float64)
     fun: 0.17269624573378842
 message: 'Optimization terminated successfully.'
     nit: 22
   slack: array([  0.        , 197.52275768,   0.19833038,  53.3585529 ,
         1.24383891,   7.46956997,   0.        ,  70.80750853,
        12.47645051])
  status: 0
 success: True
       x: array([0.12988396, 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.04281229, 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        ])

#### Interpretando o resultado da execução do linprog

O atributo fun é o preço total da dieta, correspondente a aproximadamente 17 centavos de dólar. Os itens do array x representam a quantidade de cada alimentos, como aparece na primeira tabela deste documento.
Ou seja, para o alimento farinha de trigo enriquecida, teremos uma quantidade de 0.129 unidades, aproximadamente. Para o fígado de boi 0.04281229 unidades.

O restante dos alimentos não será comprado.

#### O problema da Vitamina A pura

In [6]:
vitamina_a = tabela.loc['Vitamina A(1000I.U.)'].to_numpy()
vitamina_a = -vitamina_a

A = []
A.append(vitamina_a)
restricoes_vitamina_a = [-6.0]

res_vitamina_a = linprog(c, A_ub=A, b_ub=restricoes_vitamina_a,bounds=(0, None), method='simplex') 
res_vitamina_a.fun

0.006533101045296167

Conclui-se que para vitamina A pura, pagariamos 0.006 centavos de dólar, aproximadamente.

### Problema da riboflavina pura

In [7]:
ribo_pura = tabela.loc['Riboflavina(mg.)'].to_numpy()
ribo_pura = -ribo_pura

A = []
A.append(ribo_pura)
restricoes_ribo_pura = [-6.5]

res_ribo_pura = linprog(c, A_ub=A, b_ub=restricoes_ribo_pura,bounds=(0, None), method='simplex') 
res_ribo_pura

     con: array([], dtype=float64)
     fun: 0.1279527559055118
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([-8.8817842e-16])
  status: 0
 success: True
       x: array([0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.12795276, 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        ])

Conclui-se, portanto, que pagariamos até 0.12 centavos, aproximadamente, por riboflavina pura.